src/server/tran_server.cpp

```diff
@@ -138,6 +138,33 @@ tran_server::boot (const char *db_name)
   return NO_ERROR;
 }
 
+void
+tran_server::push_request (size_t idx, tran_to_page_request reqid, std::string &&payload)
+{
+  assert (idx < m_page_server_conn_vec.size());
+  m_page_server_conn_vec[idx]->push_request (reqid, std::move (payload));
+}
+
+int
+tran_server::send_receive (size_t idx, tran_to_page_request reqid, std::string &&payload_in,
+			   std::string &payload_out) const
+{
+  assert (idx < m_page_server_conn_vec.size());
```

Comment:
> Perhaps, I understand that this issue will simply proceed with refactoring, and the issue above (disconnected abnormally) will be dealt with later.

This is also my understanding. This refactoring should contain no functional changes.
In that regard it only deals with starting all PS's once when ATS is started and disconnecting all PS's when ATS is stopped (via dtors mainly).

Reply:
> I think above assertion means is_page_server_connected () which was used previously in push_request_to (),
and I think it's better to keep using the function as previous because it is easier to understand. 

In fact, `is_page_server_connected ()` has no meaning after refactoring. Previously, the communication interface like `push_request()` doesn't know a connection has been established.   But, now they are inside the connection, so if they can be called, it means the connection alreadt has been established.

Reply:
> Nevertheless, yes, some more handling is needed. But I'm not sure the size would be shrunk. I guess we can just let it empty, and put a new or reconnection connection to it. Let's see then.

I don't think I understand fully @joohok 's concern:

> Currently, the size of that vector does not consider the connection with the PS that was disconnected in the middle.
(the size of the m_page_server_conn_vec does not change if TS is initially connected to 10 PSs and disconnected from 2 PSs in the middle.)

But here are my 2cents.
There are a few ways in which a connection handler can break down once it is established:
- the other side (PS) does not respond anymore (crashed, hanged ...) and the send buffers in the network layer - not even in the application, but down from the OS upwards, become full and some error is returned back from somewhere in cubrid's css code; the connection handler will need a way to cope with this - take a look how it is done in page_server/connection_handler - each connection handler has a reference to the page_server object instance an can basically call upwards with "remove me: I'm dead/invalid/hanged/cannot work anymore"
- it can be removed by the owner/parent - tran_server object instance as a result of some higher-up decision (eg: cluster management software deciding that one PS is to be de-activated; or some error in the higher application layers)
- what not ...

There are also transient errors which the connection handler (+ the infrastructure below) needs to cope with (eg: like a single request fails with some error code; or with a timeout - we don't currently have timeouts implemented, but we'll need to add them).

None are implemented here with this patch.

Reply:
OK, I understand. 
Thanks for elaboration!